Implement tagged and dynamic metrics

awjchen opened the pull request:
This PR adds support for tagged and dynamic metrics, via a strongly-typed interface based on the [monad-metrics-extensible](https://hackage.haskell.org/package/monad-metrics-extensible) package.

Haddocks for the new interface may be temporarily found [here](https://hasura.github.io/temp-new-ekg-api/ekg-core/System-Metrics.html). A tutorial for this new interface has also been written -- see [Tutorial.md](https://github.com/hasura/ekg-core/blob/alex/newApi-5/Tutorial.md).

This PR also addresses issues 24, 29, 31, and 41 of tibbe/ekg-core, incorporating work from PRs 25, 30, and 42 of tibbe/ekg-core.

This PR also drops support for older versions of GHC using `base < 4.10`.

Related PRs:
- https://github.com/hasura/ekg-json/pull/2
- https://github.com/hasura/ekg/pull/1
- https://github.com/hasura/ekg-statsd/pull/2

awjchen committed:
Add a minimal sample hie.yaml

awjchen committed:
Add Prometheus labels

Allow key-value pairs to be attached to metric names and identify
metrics by both their name and key-value pairs.

awjchen committed:
Replace metrics when registering an existing identifier

When registering a metric using an identifier that is already in the
store, replace the existing metric.

Previously, "alreadyInUseError" would be thrown instead.

awjchen committed:
Allow deregistration of metrics by metric name

awjchen committed:
Initial version of statically defined metrics

awjchen committed:
Factor out pure state operations for testing

awjchen committed:
Rearrange code

awjchen committed:
Use atomicModifyIORef'

awjchen committed:
Add function to verify state consistency

awjchen committed:
Encapsulate store state in internal module

awjchen committed:
Start test suite

awjchen committed:
Fix internal state consistency issues

awjchen committed:
Add metric tags to static metrics definition

awjchen committed:
Derive instances of ToTags with GHC.Generics

awjchen committed:
Add/update documentation

awjchen committed:
Use hspec for tests

awjchen committed:
Document the test

awjchen committed:
Add a QuickCheck version of the test

awjchen committed:
Add `deregisterByName` to the static interface

awjchen committed:
Partially document the static interface

awjchen committed:
Cleanup tests

awjchen committed:
Complete static API (except for `registerGroup`)

awjchen committed:
Cleanups

awjchen committed:
Rename module

awjchen committed:
Reorganize tests

awjchen committed:
Fix min/max for the Distribution metric

I have chosen to report all statistics from the empty distribution to be
zero. This is to accommodate downstream consumers that do not parse
'NaN' or 'Infinity' as floats.

One alternative to this approach is to avoid emitting any statistics at
all for empty distributions.

awjchen committed:
Use "criterion" for benchmarks

awjchen committed:
Reimplement `Atomic` with GHC prim ops

This commit attempts to address issue #41 of tibbe/ekg-core by
replaceing the C code for `Atomic` with GHC prim ops
(`fetchAddIntArray`).

However, we insist on a 64-bit counter, so if machine does not
support 64-bit prim ops, we fall back to using an `IORef` and
`atomicModifyIORefCAS`.

The performance of the 64-bit prim ops implementation is somewhat slower
than the existing C code, perhaps due to the additional conversion
between `Int` and `Int64`.

awjchen committed:
Implement `registerGroup` for the static interface

awjchen committed:
Reorganize and document the static API

awjchen committed:
Allow restrictions of metric store scope

awjchen committed:
Relax hashable version bound

awjchen committed:
Change expected order of metric type indices

This makes the use of free-form metrics
(via `Metric :: AllMetrics name metricType tags`)
a bit more ergonomic.

Old: createCounter (Metric @_ @"total_requests")
New: createCounter (Metric @"total_requests")

awjchen committed:
Allow deregistration of metrics by identifier

awjchen committed:
Reimplement distribution metric with GHC prim ops

This commit attempts to address issue #41 of tibbe/ekg-core by
replacing the C code for the distribution metric with GHC prim ops.

The performance of this implementation is about half that of the
existing C code in a single-threaded benchmark; without masking the
performance is comparable.

This commit is based on the work of Travis Whitaker in PR #42 of
tibbe/ekg-core.

awjchen committed:
Add a test for `Distribution.addN`

awjchen committed:
Fix `Distribution.addN`

awjchen committed:
Remove unused C code

awjchen committed:
Potentially fix issue #24 of tibbe/ekg-core

Don't combine distributions when the left distribution is empty.

awjchen committed:
Ensure safety of distribution implementation

Use the "inspection-testing" library to ensure that operations on
distributions that take locks do not allocate memory.

If they did, threads running those functions could receive exceptions or
be descheduled by the runtime while holding the lock, which could result
in deadlock or severe performance degredation, respectively.

This removes the need to wrap the distribution operations in `mask`, and
thereby improves performance. The performance of this implementation is
comparable to that of the original C code in a single-threaded
benchmark; it also scales better with multiple cores.

To eliminate allocations from the distribution operations, the
operations have been rewritten entirely in terms of prim ops from
"GHC.Prim".

Warning: the "count" field of the distribution may now be limited to 31
bits on 32-bit platforms.

awjchen committed:
Tweak haddocks

awjchen committed:
Implement deregistration by "handles"

Have registration functions return IO actions ("handles") that
deregister the newly registered metrics, and only those metrics.

This is implemented by having the store track a "version" number for
each metric, which is incremented upon registration, and associating
deregistration actions to specific versions. Deregistration actions
first check whether their version matches the current version of the
target metric, and then abort if it doesn't.

awjchen committed:
Implement atomic (de)registration of multiple metrics

awjchen committed:
Move and rename modules

Making "System.Metrics.Static" (now "System.Metrics") the only exposed
interface.

awjchen committed:
Revise haddocks

awjchen committed:
Specify GC metrics using the new GADT style

Also:
- Remove support for GC metrics from base < 4.10 (July 2017, GHC 8.2.1)
- Remove deprecated alias for par_tot_bytes_copied (base < 4.6)

awjchen committed:
Add Monoid instance for `Registration`

awjchen committed:
Update empty RTSStats for base >= 4.14.1

awjchen committed:
Cabal file maintenance

- Add the standard warnings
- Remove unsupported GHCs from "tested-with"

awjchen committed:
Address compiler, hlint warnings

It's now ok not to import (<$>) from Control.Applicative because we have
dropped support for base < 4.10.0

awjchen committed:
Cleanup System.Metrics.Internal.State

awjchen committed:
Change GroupId type alias from Int to Integer

Since we can now deregister metrics, we could *technically* have an
unbounded number of groups, so we need to use `Integer` instead of `Int`
to be *completely* safe.

This also meant replacing `IntMap` with the slower `Map`. This is not a
concern for us because the only performance-sensitive operation on the
store is sampling, which acts on the group map using `elems`, and the
performance of `elems` between `IntMap`s and `Map`s are
indistinguishable (at least from a quick criterion benchmark).

awjchen committed:
More cleanups

awjchen committed:
Reorganize and extend tests

awjchen committed:
Follow new GHC.Stats naming for GC metrics

Based on the PR submitted by nh2 to tibbe/ekg-core.

awjchen committed:
Add new GHC.Stats metrics (non-moving gc metrics)

awjchen committed:
Hide the alternative deregistration interface

Hide the extra deregistration functions in order to be conservative with
the initial API. We can always re-expose the deregistration functions if
they are needed.

awjchen committed:
Add tutorial and revise Haddocks

- Moved some examples/content from the Haddocks to the tutorial

awjchen committed:
Add fallback `Distribution` for 32-bit systems

On 32-bit systems, the counter used in the implementation of
`Distribution` may be truncated to 32 bits, putting the counter at risk
of overflow. Therefore, this commit implements an alternative, fallback
implementation of `Distribution` for 32-bit systems. This alternative
implementation uses `IORef` and `atomicModifyIORefCAS`, and is much
slower.

awjchen committed:
Cleanup (renaming)

awjchen committed:
Revert changes from old fork

paf31 commented on System/Metrics.hs:
```diff
@@ -114,109 +155,308 @@ import qualified System.Metrics.Label as Label
+-- scoped to a subset of its metrics. This can be done as long as the
+-- subset can be represented by a function (see `subset`).
+
+-- | Create a new reference to a metric store with restricted scope.
```
Might be worth adding an example in the published comments along with a justification (in a non-published comment) for why `coerce` is a safe implementation.

awjchen replied:
Sounds good!

Though, I suppose this function is not safe because it never evaluates its first argument. This allows a user to change scope arbitrarily via `subset undefined`. I don't know how to prevent the use of `undefined` ... I was hoping that we could trust users to not use `undefined`, or maybe allow them to use it as an escape hatch.

paf31 replied:
Thanks for adding the example - can you please explain the intended purpose of the scopes/subset functionality though, because I'm not seeing it. Are we using it?

awjchen replied:
Sure! I had two motivations when introducing these "scopes"/"subsets".

(a) I wanted to preserve an `ekg-core` feature in the presence of the statically-defined metrics introduced in this fork. The `ekg-core` feature is that users can register to their store metrics defined by other code (i.e. libraries). Since this fork requires that all metrics be statically defined, the library code must register their metrics under some metrics specification. "Subsets" were a convenient way for users to include those library-defined metrics specifications in their own metrics specifications, thereby allowing users to register library-defined metrics. I didn't think further than this, though. Perhaps there is another solution not involving these "scopes".

In my integration of this fork with our codebase, I use "scopes" to register the pre-defined RTS metrics provided by `ekg-core`, and to register the `ServerMetrics` defined by `graphql-engine`.

(b) `ekg-core` supports static metrics only, but this fork also supports dynamic metrics (that is, metrics can be deregistered and re-registered). This support for dynamic metrics expands the number of programs that compile and run without error. For example, in `ekg-core`, registering a metric that already exists results in an exception, whereas in this fork the existing metric would be replaced, since that is now a valid operation.

Using "scopes" was a way to recover the safety of static metrics for those metrics that don't need to be -- and shouldn't be -- deregistered. To make static metrics with "scopes", you would 1. separate your dynamic and static metrics into two "subsets" of your metrics specification, 2. register the static metrics at startup, then 3. create a new reference to the store scoped only to the dynamic metrics and discard the original store reference. This statically guarantees that the metrics in the "static" subset can no longer be deregistered.

In my integration of this fork with our codebase, I used this "technique", but it is not essential.

Does all this sound reasonable to you?

awjchen replied:
I've added source code comments explaining why `subset`'s implementation is safe, and added a section to the tutorial showing how to simulate static metrics (what I tried to explain in the above part (b)).

Although the scopes/subsets may be a sub-optimal API and may need to be revised or replaced, we will have the opportunity to do so because this library will only be used internally for some time. Then, I'm going to go ahead and merge the current API with scopes/subsets with the intention of returning to it later.

awjchen committed:
Docs: Add example for `subset`

paf31 commented on Data/Atomic.hs:
```diff
@@ -12,33 +13,52 @@ module Data.Atomic
+
+-- We want an atomic with at least 64 bits in order to avoid overflow.
+
+#if WORD_SIZE_IN_BITS >= 64
```
I'm curious about the need for optimizing this so carefully here. It's worth adding a comment at least about why we need to optimize this so much. My intuition is that for our needs, this could be premature optimization, given that the time to compute some estimators is probably a drop in the ocean, given the amount of JSON serialization and network IO that we're doing in the course of reporting this stuff. I could believe it's needed though, so any elucidation in comments here would be appreciated.

awjchen replied:
Hmmmm. I don't have much knowledge or experience about optimization, so I might not be very helpful here. @jberryman, could I have your take on this?

The use of fetch-and-add here is inherited from the replaced `ekg-core` code, the implementation of which was motivated by reports of contention in `ekg` under an old implementation that used `atomicModifyIORef'` -- see the last section of [Tibell's post](https://blog.johantibell.com/2014/05/ekg-04-released.html). Indeed, repeating Tibell's contention-heavy micro-benchmark, fetch-and-add was over 100x faster than `atomicModifyIORef` and about 5x as fast as [`atomicModifyIORefCas_`](https://hackage.haskell.org/package/atomic-primops-0.8.4/docs/Data-Atomics.html#v:atomicModifyIORefCAS_). It seems highly unlikely that our use case would trigger that level of contention, but there seems also seems to be little downside to using fetch-and-add here.

Padding the counter to 64 bytes is new -- I copied that technique from `unagi-chan`. However, unlike fetch-and-add, it has the downside that it uses additional memory, so this might be premature optimization. @jberryman was also skeptical about this padding when I showed to him a while back. It would certainly do no harm to our use case if we removed this padding.

In any case, I'll add a comment.

awjchen replied:
I've added commits removing the padding for the counter metric and justifying it for the distribution metric.

jkachmar commented on System/Metrics/Distribution.hs:
```diff
@@ -0,0 +1,442 @@
+    inf :: Double
+    inf = 1/0
+
+spinlock#
+  :: MutableByteArray# RealWorld -> State# RealWorld -> State# RealWorld
+spinlock# arr s0 =
+  case posLock of { I# posLock' ->
+  case casIntArray# arr posLock' 0# 1# s0 of { (# s1, r #) ->
+  case r of
+    0# -> s1
+    _ -> spinlock# arr s1
+  }}
```
I'm always a little wary whenever I see spinlocks, but I'm not 100% sure about the constraints you're working with here so in lieu of offering targeted advice I'll defer to [this blog post by Aleksey Kladov](https://matklad.github.io/2020/01/02/spinlocks-considered-harmful.html) as well as his [follow-up indicating that "sleeping mutexes" should be faster anyway](https://matklad.github.io/2020/01/04/mutexes-are-faster-than-spinlocks.html).

awjchen replied:
Thanks for showing me the dangers of spinlocks! To be honest, I don't know what I'm doing when it comes to spinlocks or locking in general. This code is a translation of and replacement for some C code that was previously in this `ekg-core` library (and the replacement was motivated by reports that the C code was generating nonsensical results on aarch64). But now I think I might need to find an alternative locking mechanism ...

awjchen replied:
I would like to postpone replacing the spinlock. Rather than doing any more work on the `Distribution` metric type, I think we should replace it entirely, maybe with an implementation of Prometheus' histograms and summaries?

evertedsphere commented on Tutorial.md:
```diff
@@ -0,0 +1,481 @@
+used.
+
+Here is an example program that illustrates the reregistration and
+deregistrtation of metrics:
```
s/trt/tr/

awjchen replied:
:pray:

evertedsphere commented on Tutorial.md:
```diff
@@ -0,0 +1,481 @@
+        ]
+  assert (sample2 == expectedSample2) $ pure ()
+
+  -- Use the dereistration handle to deregister the original metrics.
```
s/rei/regi/

awjchen replied:
:pray:

evertedsphere commented on Tutorial.md:
```diff
@@ -0,0 +1,481 @@
+
+## Sampling subsets of metrics atomically
+
+Sampling metrics is _not_ atomic; however, subsets of metrics _can_ be
```
it'd help to rewrite the first clause to explain that sampling multiple metrics individually in sequence is not guaranteed to be atomic — atomicity or a lack thereof isn't defined in the context of sampling a single primitive value, right?

awjchen replied:
This is true, though ... I think I completely botched this sentence :sweat_smile: It doesn't really communicate the point I wanted to make, so I've rewritten it now.

evertedsphere replied:
ah, sampleAll is not atomic, that makes sense (because it's multiple metrics)

evertedsphere replied:
why not make `sampleAll` atomic by default (if that's possible)?

evertedsphere replied:
or, well, rename it to `sampleAllNonAtomic` or something, and recommend using an atomic `sampleAll` function by default unless the user is happy to be provided a possibly-internally-inconsistent snapshot of the metrics in which they can use the non-atomic function instead

awjchen replied:
Hmmmm, I think that, if the sampling actions we accept from the users can be arbitrary `IO` actions, it's not possible for us to guarantee the atomicity of a sample containing multiple metrics.

Even when we have an atomically-sampled group of metrics registered via `registerGroup`, we are relying on the user to give us an `IO` action that atomically produces the group's shared value. In that sense, the `registerGroup` function just allows a user to express a value that _they_ compute atomically in a format that `ekg-core` can understand (i.e. by projecting that value onto multiple metrics).

evertedsphere replied:
ah i get it now. i think the docs could be improved to explain this better, just the way you did above, perhaps?

awjchen replied:
I've updated the tutorial to include the bit about why sampling isn't atomic!

evertedsphere commented on Tutorial.md:
```diff
@@ -0,0 +1,481 @@
+
+`ekg-core` has a multi-dimensional data model, like
+[Prometheus](https://prometheus.io). In this data model, metrics may be
+annotated by sets of key-value pairs called **tags** or **tag sets**.
```
perhaps
```diff
- metrics may be annotated by sets of key-value pairs called **tags** or **tag sets**.
+ metrics may be annotated by a **tag set**, which is a set 
+ of key-value pairs called **tags**.
```

awjchen replied:
This is great!

evertedsphere replied:
:)

evertedsphere reviewed (approved):
looks good to me! 

love the direction you chose when designing the new api. outside of that, the store action tests are also really cool, and the tutorial is quite well-written too

(left some nitpicks over typos etc, feel free to delete them or mark them resolved once you've gone over them)

awjchen committed:
Revise tutorial

awjchen committed:
Tutorial: Elaborate on atomic sampling of metrics

awjchen committed:
Tutorial: Show how to simulate static metrics

awjchen committed:
Remove cache line padding from Data.Atomic

Padding to 64 bytes (an x86 cache line) uses 8x the memory but probably
only matters for the most demanding use cases.

awjchen committed:
Justify cache line padding for the distribution metric

awjchen committed:
Tutorial: Revise section on atomic sampling

awjchen committed:
Add tests for thread-safety

... by repurposing the contention-heavy benchmarks

awjchen committed:
Explain the `subset` function in source code comments

awjchen commented:
This PR now also has approval from @jberryman. Merging now. Also merging the other PRs.
